File: Design and Analysis of Algorithms/police_cars.cpp
// C++ program for finding minimum cut using Ford-Fulkerson
#include <iostream>
#include <limits.h>
#include <string.h>
#include <queue>
using namespace std;

// Number of vertices in given graph

#define V 10000000

long long int answer = 0;
int n, m, a, b;

/* Returns true if there is a path from source 's' to sink 't' in
residual graph. Also fills parent[] to store the path */
int bfs(int rGraph[V][V], int s, int t, int parent[])
{
	// Create a visited array and mark all vertices as not visited
	bool visited[n];
	memset(visited, 0, sizeof(visited));

	// Create a queue, enqueue source vertex and mark source vertex
	// as visited
	queue <int> q;
	q.push(s);
	visited[s] = true;
	parent[s] = -1;

	// Standard BFS Loop
	while (!q.empty())
	{
		int u = q.front();
		q.pop();

		for (int v=0; v<n; v++)
		{
			if (visited[v]==false && rGraph[u][v] > 0)
			{
				q.push(v);
				parent[v] = u;
				visited[v] = true;
			}
		}
	}

	// If we reached sink in BFS starting from source, then return
	// true, else false
	return (visited[t] == true);
}

// A DFS based function to find all reachable vertices from s. The function
// marks visited[i] as true if i is reachable from s. The initial values in
// visited[] must be false. We can also use BFS to find reachable vertices
void dfs(int rGraph[V][V], int s, bool visited[])
{
	visited[s] = true;
	for (int i = 0; i < n; i++)
    {
        if (rGraph[s][i] && !visited[i])
        {
            dfs(rGraph, i, visited);
        }
    }
}

// Prints the minimum s-t cut
void minCut(int **graph, int s, int t)
{
	int u, v;

	// Create a residual graph and fill the residual graph with
	// given capacities in the original graph as residual capacities
	// in residual graph
	int rGraph[n][n]; // rGraph[i][j] indicates residual capacity of edge i-j
	for (u = 0; u < n; u++)
    {
        for (v = 0; v < n; v++)
        {
            rGraph[u][v] = graph[u][v];
        }
    }

	int parent[n]; // This array is filled by BFS and to store path

	// Augment the flow while there is a path from source to sink
	// while (bfs(rGraph, s, t, parent))
	// {
	// 	// Find minimum residual capacity of the edges along the
	// 	// path filled by BFS. Or we can say find the maximum flow
	// 	// through the path found.
	// 	int path_flow = INT_MAX;
	// 	for (v=t; v!=s; v=parent[v])
	// 	{
	// 		u = parent[v];
	// 		path_flow = min(path_flow, rGraph[u][v]);
	// 	}

	// 	// update residual capacities of the edges and reverse edges
	// 	// along the path
	// 	for (v=t; v != s; v=parent[v])
	// 	{
	// 		u = parent[v];
	// 		rGraph[u][v] -= path_flow;
	// 		rGraph[v][u] += path_flow;
	// 	}
	// }

	// Flow is maximum now, find vertices reachable from s
	bool visited[n];
	memset(visited, false, sizeof(visited));
	// dfs(rGraph, s, visited);

	// Print all edges that are from a reachable vertex to
	// non-reachable vertex in the original graph
	for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < n; j++)
        {
            if(visited[i] && !visited[j] && graph[i][j])
            {
                // cout << i << " - " << j << endl;
                answer = answer + graph[i][j];
            }
        }
    }
	return;
}

// Driver program to test above functions
int main()
{
    scanf("%d %d %d %d", &n, &m, &a, &b);
	// Let us create a graph shown in the above example
    int **graph;
    for(int i=0; i<m; i++)
    {
        int l, r, c;
        scanf("%d %d %d", &l, &r, &c);
        graph[l-1][r-1] = c;
        graph[r-1][l-1] = c;
    }

	minCut(graph, a-1, b-1);

	return 0;
}

File: Design and Analysis of Algorithms/police_cars.py
# Python program for finding min-cut in the given graph 
# Complexity : (E*(V^3)) 
# Total augmenting path = VE and BFS 
# with adj matrix takes :V^2 times 

from collections import defaultdict
import sys
# This class represents a directed graph
# using adjacency matrix representation

# answer = 0
class Graph: 

	def __init__(self,graph): 
		self.graph = graph # residual graph 
		self.org_graph = [i[:] for i in graph] 
		self. ROW = len(graph) 
		self.COL = len(graph[0]) 
		self.answer = 0


	'''Returns true if there is a path from 
	source 's' to sink 't' in 
	residual graph. Also fills 
	parent[] to store the path '''
	def BFS(self,s, t, parent): 

		# Mark all the vertices as not visited 
		visited =[False]*(self.ROW) 

		# Create a queue for BFS 
		queue=[] 

		# Mark the source node as visited and enqueue it 
		queue.append(s) 
		visited[s] = True

		# Standard BFS Loop 
		while queue: 

			#Dequeue a vertex from queue and print it 
			u = queue.pop(0) 

			# Get all adjacent vertices of 
			# the dequeued vertex u 
			# If a adjacent has not been
			# visited, then mark it 
			# visited and enqueue it 
			for ind, val in enumerate(self.graph[u]): 
				if visited[ind] == False and val > 0 : 
					queue.append(ind) 
					visited[ind] = True
					parent[ind] = u 

		# If we reached sink in BFS starting
		# from source, then return 
		# true, else false 
		return True if visited[t] else False
		
	# Function for Depth first search 
	# Traversal of the graph
	def dfs(self, graph,s,visited):
		visited[s]=True
		for i in range(len(graph)):
			if graph[s][i]>0 and not visited[i]:
				self.dfs(graph,i,visited)

	# Returns the min-cut of the given graph 
	def minCut(self, source, sink): 

		# This array is filled by BFS and to store path 
		parent = [-1]*(self.ROW) 

		max_flow = 0 # There is no flow initially 

		# Augment the flow while there is path from source to sink 
		while self.BFS(source, sink, parent) : 

			# Find minimum residual capacity of the edges along the 
			# path filled by BFS. Or we can say find the maximum flow 
			# through the path found. 
			path_flow = float("Inf") 
			s = sink 
			while(s != source): 
				path_flow = min (path_flow, self.graph[parent[s]][s]) 
				s = parent[s] 

			# Add path flow to overall flow 
			max_flow += path_flow 

			# update residual capacities of the edges and reverse edges 
			# along the path 
			v = sink 
			while(v != source): 
				u = parent[v] 
				self.graph[u][v] -= path_flow 
				self.graph[v][u] += path_flow 
				v = parent[v] 

		visited=len(self.graph)*[False]
		self.dfs(self.graph,s,visited)

		# print the edges which initially had weights 
		# but now have 0 weight 
		for i in range(self.ROW): 
			for j in range(self.COL): 
				if self.graph[i][j] == 0 and self.org_graph[i][j] > 0 and visited[i]: 
					# print(str(i) + " - " + str(j))
					self.answer = self.org_graph[i][j] + self.answer
		return self.answer


# Create a graph given in the above diagram
elem = input().split(' ')
n = int(elem[0])
m = int(elem[1])
a = int(elem[2])
b = int(elem[3])
# print(a, b, m, n)
graph = [[0 for x in range(n)] for y in range(n)]

sys.setrecursionlimit(n)


for i in range(m):
	edge = input().split(' ')
	c = int(edge[0])
	d = int(edge[1])
	w = int(edge[2])
	# print(c, d, w, " is c d w")
	graph[c-1][d-1] = w
	graph[d-1][c-1] = w

# print(graph)
g = Graph(graph)

source = a-1; sink = b-1

print(g.minCut(source, sink))

# This code is contributed by Neelam Yadav 
